shared/subspace-data-retrieval/src/segment_fetcher.rs

```diff
@@ -85,66 +80,87 @@ where
 {
     debug!(%segment_index, "Retrieving pieces of the segment");
 
-    let semaphore = &Semaphore::new(RecordedHistorySegment::NUM_RAW_RECORDS);
-
-    let mut received_segment_pieces = segment_index
-        .segment_piece_indexes_source_first()
-        .into_iter()
-        .map(|piece_index| {
-            // Source pieces will acquire permit here right away
-            let maybe_permit = semaphore.try_acquire();
-
-            async move {
-                let permit = match maybe_permit {
-                    Some(permit) => permit,
-                    None => {
-                        // Other pieces will acquire permit here instead
-                        semaphore.acquire().await
-                    }
-                };
-                let piece = match piece_getter.get_piece(piece_index).await {
-                    Ok(Some(piece)) => piece,
-                    Ok(None) => {
-                        trace!(?piece_index, "Piece not found");
-                        return None;
-                    }
-                    Err(error) => {
-                        trace!(
-                            %error,
-                            ?piece_index,
-                            "Piece request failed",
-                        );
-                        return None;
-                    }
-                };
-
-                trace!(?piece_index, "Piece request succeeded");
-
-                // Piece was received successfully, "remove" this slot from semaphore
-                permit.forget();
-                Some((piece_index, piece))
-            }
-        })
-        .collect::<FuturesUnordered<_>>();
-
+    // We want NUM_RAW_RECORDS pieces to reconstruct the segment, but it doesn't matter exactly which ones.
+    let piece_indexes = segment_index.segment_piece_indexes_source_first();
+    let mut piece_indexes = piece_indexes.as_slice();
     let mut segment_pieces = vec![None::<Piece>; ArchivedHistorySegment::NUM_PIECES];
+
+    let mut pieces_pending = 0;
     let mut pieces_received = 0;
+    let mut piece_streams = Vec::new();
+
+    // Loop Invariant:
+    // - the number of remaining piece indexes gets smaller, eventually finishing the fetcher, or
+    // - the number of pending pieces gets smaller, eventually triggering another batch.
+    // We also exit early if we have enough pieces to reconstruct a segment.
+    'fetcher: while !piece_indexes.is_empty()
```

Comment:
I would split this into two stages:
* first try to download all 128 source pieces, if successful reconstruction will be very cheap and fast
* as a fallback when actual reconstruction is needed, schedule more pieces to download, including parity

Right now it is implemented in a way that is a bit wasteful in terms of bandwidth (triggers more downloads than needed) and in terms of CPU usage (has overwhelmingly high chance of not getting 128 source pieces for cheap segment reconstruction).

Reply:
This is already what the code does?

The first batch contains 128 source piece indexes. If all pieces in that batch succeed, then there aren’t any parity piece requests.

But as soon as any pieces fail, a batch of parity piece indexes is created, which contains exactly the number of pieces needed to compensate for those failures. Then all batches are polled concurrently.

The code assumes that any pieces that are still pending will succeed, so there’s no wasted downloads.

Reply:
I see, the way it is written wasn't as clear, but now I understand what it does. If you exhaust the stream of pieces you've generated every time, why do you keep already finished streams in `piece_streams` (the question above)? I don't see how multiple streams can be pooled concurrently here.

Reply:
> I see, the way it is written wasn't as clear, but now I understand what it does.

Good feedback, I might split it into multiple methods so it's clearer.

`flatten_unordered()` polls all the streams in the vector concurrently, and can return the pieces from any stream.

`ready_chunks()` waits until at least one piece result is ready, then returns all the ready pieces as a vector. But if any pieces are still pending, they are left in the stream.

Then if any of the piece results in that vector are `None`, we add a batch of parity pieces to replace them.

We can definitely drop streams once they're done, I'll make some notes about how to do that.

Reply:
I actually already wrote an efficient piece retrieval for reconstruction purposes in https://github.com/autonomys/subspace/blob/362c1f5dce076b6c13452977a533f9091d515bb1/crates/subspace-farmer-components/src/segment_reconstruction.rs

It might be a little simpler and it does try to download pieces in batches all the time, avoiding batches of 1 that you will most likely get majority of time after initial request due to the way `ready_chunks` works. It is overall less eager and tries to not do a lot of heavy lifting.

Can be extracted into a utility somewhere to return a segment worth of pieces and then reused in farmer code, here and in DSN sync on the node, where exactly the same logic will be necessary for piece retrieval, just what we do with those pieces is slightly different.

Reply:
I think I'll do this in another PR